mm/mm_heap/mm_free.c

```diff
@@ -78,34 +78,37 @@ static void add_delaylist(FAR struct mm_heap_s *heap, FAR void *mem)
  *
  ****************************************************************************/
 
-void mm_delayfree(FAR struct mm_heap_s *heap, FAR void *mem, bool delay)
+void mm_delayfree(FAR struct mm_heap_s *heap, FAR void *mem,
+                  bool delay, bool cleanup)
 {
   FAR struct mm_freenode_s *node;
   FAR struct mm_freenode_s *prev;
   FAR struct mm_freenode_s *next;
   size_t nodesize;
   size_t prevsize;
 
-  if (mm_lock(heap) < 0)
+  if (cleanup)
     {
-      /* Meet -ESRCH return, which means we are in situations
-       * during context switching(See mm_lock() & gettid()).
-       * Then add to the delay list.
-       */
-
-      add_delaylist(heap, mem);
-      return;
-    }
-
 #ifdef CONFIG_MM_FILL_ALLOCATIONS
```

Comment:
both CONFIG_MM_FILL_ALLOCATIONS and kasan_poison is for debugging purpose, why do we need optimize the debugging path  but compilated the code logic?

Reply:
I just found these potential issues when reading the code:
1. memset/kasan is executed in the scope of mm_lock(), these operations will reduce the allocator usage in multi-thread case.
2. memset/kasan will be executed twice in delay free case, which is unreasonable

Reply:
Do you think system performance is unimportant with debugging feature enabled? If so, I will close this PR

Reply:
Ok, let's continue this patch